File: FBControlCore/Utility/FBControlCoreError.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBControlCoreError.h"

#import "FBControlCoreGlobalConfiguration.h"
#import "FBControlCoreLogger.h"
#import "FBFuture.h"

NSString *const FBControlCoreErrorDomain = @"com.facebook.FBControlCore";

@interface FBControlCoreError ()

@property (nonatomic, readwrite, copy) NSString *domain;
@property (nonatomic, readwrite, copy) NSString *describedAs;
@property (nonatomic, readwrite, copy) NSError *cause;
@property (nonatomic, readwrite, strong) NSMutableDictionary *additionalInfo;
@property (nonatomic, readwrite, assign) BOOL describeRecursively;
@property (nonatomic, readwrite, assign) NSInteger code;

@end

@implementation FBControlCoreError

#pragma mark Initializers

- (instancetype)init
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _domain = FBControlCoreErrorDomain;
  _code = 0;
  _additionalInfo = [NSMutableDictionary dictionary];
  _describeRecursively = YES;

  return self;
}

#pragma mark Public Methods

+ (instancetype)describe:(NSString *)description
{
  return [self.new describe:description];
}

- (instancetype)describe:(NSString *)description
{
  self.describedAs = description;
  return self;
}

+ (instancetype)causedBy:(NSError *)cause
{
  return [self.new causedBy:cause];
}

- (instancetype)causedBy:(NSError *)cause
{
  self.cause = cause;
  return self;
}

- (BOOL)failBool:(NSError **)error
{
  if (error) {
    *error = [self build];
  }
  return NO;
}

- (int)failInt:(NSError **)error
{
  if (error) {
    *error = [self build];
  }
  return 0;
}

- (unsigned int)failUInt:(NSError **)error
{
  if (error) {
    *error = [self build];
  }
  return 0;
}

- (CGRect)failRect:(NSError **)error
{
  if (error) {
    *error = [self build];
  }
  return CGRectNull;
}

- (id)fail:(NSError **)error
{
  if (error) {
    *error = [self build];
  }
  return nil;
}

- (FBFuture *)failFuture
{
  NSError *error = [self build];
  return [FBFuture futureWithError:error];
}

- (FBFutureContext *)failFutureContext
{
  return [FBFutureContext futureContextWithError:[self build]];
}

- (void *)failPointer:(NSError **)error;
{
  if (error) {
    *error = [self build];
  }
  return NULL;
}

- (instancetype)extraInfo:(NSString *)key value:(id)value
{
  if (!key || !value) {
    return self;
  }
  self.additionalInfo[key] = value;
  return self;
}

- (instancetype)recursiveDescription
{
  self.describeRecursively = YES;
  return self;
}

- (instancetype)noRecursiveDescription
{
  self.describeRecursively = NO;
  return self;
}

- (instancetype)inDomain:(NSString *)domain
{
  self.domain = domain;
  return self;
}

- (instancetype)code:(NSInteger)code
{
  self.code = code;
  return self;
}

- (NSError *)build
{
  // If there's just a cause, there's no error to build
  if (self.cause && !self.describedAs && self.additionalInfo.count == 0) {
    return self.cause;
  }

  NSMutableDictionary *userInfo = [NSMutableDictionary dictionary];
  if (self.describedAs) {
    userInfo[NSLocalizedDescriptionKey] = self.describedAs;
  }
  if (self.cause) {
    userInfo[NSUnderlyingErrorKey] = self.underlyingError;
  }
  [userInfo addEntriesFromDictionary:self.additionalInfo];

  return [NSError errorWithDomain:self.domain code:self.code userInfo:[userInfo copy]];
}

#pragma mark NSObject

- (NSString *)description
{
  return [self.build description];
}

#pragma mark Private

- (NSError *)underlyingError
{
  NSError *error = self.cause;
  if (!self.describeRecursively) {
    return error;
  }
  NSError *cause = self.cause;
  if (!cause) {
    return error;
  }

  NSMutableString *description = [NSMutableString stringWithFormat:@"%@", error.localizedDescription];
  while (error.userInfo[NSUnderlyingErrorKey]) {
    error = error.userInfo[NSUnderlyingErrorKey];
    [description appendFormat:@"\nCaused By: %@", error.localizedDescription];
  }

  NSMutableDictionary *userInfo = [cause.userInfo mutableCopy];
  userInfo[NSLocalizedDescriptionKey] = description;
  return [NSError errorWithDomain:cause.domain code:cause.code userInfo:[userInfo copy]];
}

@end

@implementation FBControlCoreError (Constructors)

+ (NSError *)errorForDescription:(NSString *)description
{
  return [[self describe:description] build];
}

+ (id)failWithErrorMessage:(NSString *)errorMessage errorOut:(NSError **)errorOut
{
  return [[self describe:errorMessage] fail:errorOut];
}

+ (id)failWithError:(NSError *)failureCause errorOut:(NSError **)errorOut
{
  return [[self causedBy:failureCause] fail:errorOut];
}

+ (id)failWithError:(NSError *)failureCause description:(NSString *)description errorOut:(NSError **)errorOut
{
  return [[[self causedBy:failureCause] describe:description] fail:errorOut];
}

+ (BOOL)failBoolWithErrorMessage:(NSString *)errorMessage errorOut:(NSError **)errorOut
{
  return [[self describe:errorMessage] failBool:errorOut];
}

+ (BOOL)failBoolWithError:(NSError *)failureCause errorOut:(NSError **)errorOut
{
  return [[self causedBy:failureCause] failBool:errorOut];
}

+ (BOOL)failBoolWithError:(NSError *)failureCause description:(NSString *)description errorOut:(NSError **)errorOut
{
  return [[[self causedBy:failureCause] describe:description] failBool:errorOut];
}

+ (FBFuture *)failFutureWithError:(NSError *)error
{
  return [FBFuture futureWithError:error];
}

@end

File: FBControlCore/Utility/FBControlCoreError.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation

public let FBControlCoreErrorDomain = "com.facebook.FBControlCore"

@objc
open class FBControlCoreError: NSObject {

  // MARK: Properties

  private var domain: String
  private var describedAs: String?
  private var cause: NSError?
  private var additionalInfo: [String: Any]
  private var describeRecursively: Bool
  private var errorCode: Int

  // MARK: Initializers

  public required override init() {
    domain = FBControlCoreErrorDomain
    errorCode = 0
    additionalInfo = [:]
    describeRecursively = true
    super.init()
  }

  // MARK: Public Methods

  @objc open class func describe(_ description: String) -> Self {
    let instance = self.init()
    instance.describedAs = description
    return instance
  }

  @objc @discardableResult
  open func describe(_ description: String) -> Self {
    describedAs = description
    return self
  }

  @objc open class func caused(by cause: Error?) -> Self {
    let instance = self.init()
    instance.cause = cause.map { $0 as NSError }
    return instance
  }

  @objc @discardableResult
  open func caused(by cause: Error?) -> Self {
    self.cause = cause.map { $0 as NSError }
    return self
  }

  @objc @discardableResult
  open func failBool(_ error: NSErrorPointer) -> Bool {
    error?.pointee = build()
    return false
  }

  @objc @discardableResult
  open func failInt(_ error: NSErrorPointer) -> Int32 {
    error?.pointee = build()
    return 0
  }

  @objc @discardableResult
  open func failUInt(_ error: NSErrorPointer) -> UInt32 {
    error?.pointee = build()
    return 0
  }

  @objc @discardableResult
  open func failRect(_ error: NSErrorPointer) -> CGRect {
    error?.pointee = build()
    return CGRect.null
  }

  @objc @discardableResult
  open func fail(_ error: NSErrorPointer) -> Any? {
    error?.pointee = build()
    return nil
  }

  @objc open func failFuture() -> FBFuture<AnyObject> {
    return FBFuture<AnyObject>(error: build())
  }

  @objc open func failFutureContext() -> FBFutureContext<AnyObject> {
    return FBFutureContext<AnyObject>(error: build())
  }

  @objc @discardableResult
  open func failPointer(_ error: NSErrorPointer) -> UnsafeMutableRawPointer? {
    error?.pointee = build()
    return nil
  }

  @objc @discardableResult
  open func extraInfo(_ key: String, value: Any) -> Self {
    additionalInfo[key] = value
    return self
  }

  @objc @discardableResult
  open func recursiveDescription() -> Self {
    describeRecursively = true
    return self
  }

  @objc @discardableResult
  open func noRecursiveDescription() -> Self {
    describeRecursively = false
    return self
  }

  @objc @discardableResult
  open func inDomain(_ domain: String) -> Self {
    self.domain = domain
    return self
  }

  @objc @discardableResult
  open func code(_ code: Int) -> Self {
    errorCode = code
    return self
  }

  @objc open func build() -> NSError {
    // If there's just a cause, there's no error to build
    if let cause, describedAs == nil, additionalInfo.isEmpty {
      return cause
    }

    var userInfo: [String: Any] = [:]
    if let describedAs {
      userInfo[NSLocalizedDescriptionKey] = describedAs
    }
    if cause != nil {
      userInfo[NSUnderlyingErrorKey] = underlyingError()
    }
    for (key, value) in additionalInfo {
      userInfo[key] = value
    }

    return NSError(domain: domain, code: errorCode, userInfo: userInfo)
  }

  // MARK: NSObject

  open override var description: String {
    return build().description
  }

  // MARK: Private

  private func underlyingError() -> NSError? {
    guard let error = cause else {
      return cause
    }
    if !describeRecursively {
      return error
    }

    let description = NSMutableString(string: "\(error.localizedDescription)")
    var currentError: NSError = error
    while let underlying = currentError.userInfo[NSUnderlyingErrorKey] as? NSError {
      currentError = underlying
      description.append("\nCaused By: \(currentError.localizedDescription)")
    }

    var userInfo = error.userInfo
    userInfo[NSLocalizedDescriptionKey] = description as String
    return NSError(domain: error.domain, code: error.code, userInfo: userInfo)
  }
}

// MARK: - Constructors

extension FBControlCoreError {

  @objc public class func error(forDescription description: String) -> NSError {
    return Self.describe(description).build()
  }

  @objc public class func fail(withErrorMessage errorMessage: String, errorOut: NSErrorPointer) -> Any? {
    return Self.describe(errorMessage).fail(errorOut)
  }

  @objc(failWithError:errorOut:)
  public class func fail(with failureCause: NSError, errorOut: NSErrorPointer) -> Any? {
    return Self.caused(by: failureCause).fail(errorOut)
  }

  @objc(failWithError:description:errorOut:)
  public class func fail(with failureCause: NSError, description: String, errorOut: NSErrorPointer) -> Any? {
    return Self.caused(by: failureCause).describe(description).fail(errorOut)
  }

  @objc public class func failBool(withErrorMessage errorMessage: String, errorOut: NSErrorPointer) -> Bool {
    return Self.describe(errorMessage).failBool(errorOut)
  }

  @objc(failBoolWithError:errorOut:)
  @discardableResult
  public class func failBool(with failureCause: NSError, errorOut: NSErrorPointer) -> Bool {
    return Self.caused(by: failureCause).failBool(errorOut)
  }

  @objc(failBoolWithError:description:errorOut:)
  @discardableResult
  public class func failBool(with failureCause: NSError, description: String, errorOut: NSErrorPointer) -> Bool {
    return Self.caused(by: failureCause).describe(description).failBool(errorOut)
  }

  @objc(failFutureWithError:)
  public class func failFuture(with error: NSError) -> FBFuture<AnyObject> {
    return FBFuture<AnyObject>(error: error)
  }
}
